src/Workspaces/CoreTest/SemanticModelReuse/SemanticModelReuseTests.cs

```diff
@@ -0,0 +1,368 @@
+            var model = await document.ReuseExistingSpeculativeModelAsync(null, CancellationToken.None);
+            Assert.False(model.IsSpeculativeSemanticModel);
+        }
+
+        [Fact]
+        public async Task NullBodyReturnsNormalSemanticModel2_CSharp()
+        {
+            var source = "class C { void M() { return; } }";
+            var document = CreateDocument(source, LanguageNames.CSharp);
+
+            // Even if we've primed things with a real location, getting a semantic model for null should return a
+            // non-speculative model.
+            var model1 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            var model2 = await document.ReuseExistingSpeculativeModelAsync(null, CancellationToken.None);
+            Assert.False(model1.IsSpeculativeSemanticModel);
+            Assert.False(model2.IsSpeculativeSemanticModel);
+        }
+
+        [Fact]
+        public async Task SameSyntaxTreeReturnsNonSpeculativeModel_CSharp()
+        {
+            var source = "class C { void M() { return; } }";
+            var document = CreateDocument(source, LanguageNames.CSharp);
+
+            // First call will prime the cache to point at the real semantic model.  The next call will also use the
+            // same syntax tree, so it should get the same semantic model.
+            var model1 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            var model2 = await document.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            Assert.False(model1.IsSpeculativeSemanticModel);
+            Assert.False(model2.IsSpeculativeSemanticModel);
+
+            // Should be the same models.
+            Assert.Equal(model1, model2);
+
+            // Which also should be the normal model the document provides.
+            var actualModel = await document.GetSemanticModelAsync();
+            Assert.Equal(model1, actualModel);
+        }
+
+        [Fact]
+        public async Task InBodyEditShouldProduceCachedModel_CSharp()
+        {
+            var source = "class C { void M() { return; } }";
+            var document1 = CreateDocument(source, LanguageNames.CSharp);
+
+            // First call will prime the cache to point at the real semantic model.
+            var model1 = await document1.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            Assert.False(model1.IsSpeculativeSemanticModel);
+
+            var document2 = document1.WithText(SourceText.From("class C { void M() { return null; } }"));
+
+            // This should be able to get a speculative model using the original model we primed the cache with.
+            var model2 = await document2.ReuseExistingSpeculativeModelAsync(source.IndexOf("return"), CancellationToken.None);
+            Assert.True(model2.IsSpeculativeSemanticModel);
```

Comment:
There is ParentModel if you want to assert it's parented by the prior one, although I can imagine that might be a bit too much "asserting implementation details" if you want to stay away from that.

Reply:
Also worth adding an assert that model2.SyntaxTree.ToString() gets you in-body-edited text back, just to make sure we didn't make a speculative model but it's somehow speculating on the wrong thing?